### ISSUE TITLE
Cloud Run for Anthos

### ISSUE THREAD

**Issue by pratik141** [terraform-providers/terraform-provider-google#5054]
Is Cloud Run for Anthos is supported by Terraform?

if yes Please share the document and reference links.

If you have a support request or question please submit them to one of these resources:

* [Terraform community resources](https://www.terraform.io/docs/extend/community/index.html)
* [HashiCorp support](https://support.hashicorp.com) (Terraform Enterprise customers)
* [Google Cloud Platform Slack](https://gcp-slack.appspot.com/)

**Community Note**

Please vote on this issue by adding a 👍 reaction to the original issue to help the community and maintainers prioritize this request
Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

**Comment by chrisst** [comment]
I'm pretty sure that you are looking for what was formally known as Cloud Run on GKE, which is supported already on the `google_container_cluster` resource: https://www.terraform.io/docs/providers/google/r/container_cluster.html#cloudrun_config but only in the Google Beta provider.

If you follow the directions on the Cloud Run for Anthos page to [set up a new cluster with Cloud Run](https://cloud.google.com/run/docs/gke/setup#create_cluster) you'll see that the process to set up Cloud Run for Anthos in the cloud console is to enable Cloud Run on an existing or new GKE cluster.

For a complete example check out [the cloudrun GKE integration test](https://github.com/terraform-providers/terraform-provider-google-beta/blob/c6d289c532cd1d100acb02037d6bc5672655974e/google-beta/resource_container_cluster_test.go#L1837).

Let me know if I'm misunderstanding and you are looking for something different.

**Comment by AndrewZurn** [comment]
I was looking for what I believe is what the original author might have been looking for, that being, how to select Cloud Run for Anthos as the Deployment platform for the containers that Cloud Run will be managing. Here is screenshot for the relevant functionality from the GCP Cloud Run UI.

![image](https://user-images.githubusercontent.com/2126119/72311734-807e8600-363a-11ea-9f02-794526f4ef9e.png)

I haven't look around too much, but from an initial glance it doesn't look like there is an parameter describing this functionality in the KNative Service creation spec, so maybe it's not supported yet? Although if it's support from the UI I would assume there would be a parameter, documented or not, that allows one to specify the GKE cluster to deploy against.
Relevant Spec (I think): https://github.com/knative/docs/blob/master/docs/serving/spec/knative-api-specification-1.0.md#revisiontemplatespec

**Comment by chrisst** [comment]
@AndrewZurn The dropdown you see is filtering the list of available GKE instances to only those that have cloud run enabled. This isn't something that would be part of KNative.

It looks like the UI filter is checking for any GKE cluster that satisfies the requirements for it to be able to work with cloud run. I don't know the exact filter but it appears to more than just setting `cloudrun_config`. I suspect it's all the additional features that are enabled as per this dialog:
![Screen Shot 2020-01-22 at 7 33 11 PM](https://user-images.githubusercontent.com/1569002/72954747-46099d00-3d4e-11ea-8dd6-77e56d17756a.png)

FIWI I was able to set up a GKE cluster that showed up in the dropdown with the following config:
```
resource "google_container_cluster" "tf" {
  name               = "tf2-cloud-run-anthos"
  location           = "us-central1-a"
  initial_node_count = 0
  provider           = google-beta
  network            = "default"

  ip_allocation_policy {
    cluster_ipv4_cidr_block = "10.32.0.0/14"
    services_ipv4_cidr_block = "10.0.0.0/20"
  }

  addons_config {
    http_load_balancing {
      disabled = false
    }
    horizontal_pod_autoscaling {
      disabled = false
    }
    network_policy_config {
      disabled = true
    }
    cloudrun_config {
      disabled = false
    }
  }

  network_policy {
    enabled = false
  }


  node_pool {
    initial_node_count = 3
    node_config {
      disk_size_gb = "100"
      disk_type    = "pd-standard"
      machine_type = "n1-standard-2"
    }
  }
}
```

Closing this as resolved. Let me know if there is something I've missed.

**Comment by AndrewZurn** [comment]
I more so curious, if upon creating a cluster with cloud run enabled, that you can then specify it as the backing/deployment platform for the a new cloud run resource.

Something like this:
```
resource "google_cloud_run_service" "example-cloud-run" {
  name     = "example-tf-poc"
  location = "us-central1"
  project     = var.gcp_project_id
  deployment_platform = google_container_cluster.tf.name/id/some_identifier

  template {
    spec {
      containers {
        image = "<my_image>"
      }
    }
  }

  traffic {
    percent         = 100
    latest_revision = true
  }
}
```

Does that make sense, or am I totally in left field? More or less, we can create a GKE cluster with cloud run enabled, but how is it that we can then leverage that same cluster to deploy our cloud run services to from terraform?

**Comment by chrisst** [comment]
Ok I think what you are looking for is described a bit better in the follow documents. Setting up a GKE cluster to be equipped to be used by cloud run is described much better in [this set up guide](https://cloud.google.com/run/docs/gke/setup#create_cluster) than in my previous comment. 
To answer your question, deploying a cloud run service to an existing GKE instance is described at https://cloud.google.com/run/docs/deploying#service and can be done via gcloud. I am not sure that Terraform is a good medium for managing deployments of containers as it's not great to describe declaratively. If you have a use case that that you think is worth modelling declaratively then can you open a new issue that we can label as a new resource request?

**Comment by AndrewZurn** [comment]
@chrisst - Thanks for sharing those. Yeah, definitely sounds like opening a new issue for a new potential feature request is a good route. I'll put together some notes around what/where I see myself using this, and get that opened sometime in the next few days. Thanks again for the assistance!

**Comment by geekflyer** [comment]
@chrisst I'm just having the same issue. I want to deploy cloud run services to a GKE / Anthos Cluster.
There is no new resource required - the resource is already there: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_cloud_run_service.go  

However it currently doesn't expose the deployment platform parameter  (equivalent to `gcloud run deploy --platform=gke`). So this is just an enhancement request to the existing cloud run service resources.

I also would disagree that modeling deployments of containers cannot be described declaratively. Isn't that the whole idea of kubernetes basically?

**Comment by chrisst** [comment]
@geekflyer `gcloud run deploy` is hitting different API's depending on what is configured in the `--platform` flag. When using `platform=gke` it will query a list of GKE clusters and then apply the Knative Service configuration to the kubernetes cluster. You can add `--log-http` to your gcloud commands to see what I mean. Unfortunately gcloud and the user docs for Cloud Run don't make this distinction very clear.

Alternately the fully managed Cloud Run API is a Google Cloud API that accepts a knative _like_ configuration and then does the Kubernetes apply behind the API surface. 

The two resources share a similar schema but they are backed by fundamentally different APIs. To support sending requests to a Kubernetes cluster it would be important to support all of the auth patterns that are supported by Kubernetes so that a resource could be used on any user's cluster. Whereas supporting the cloud APIs requires a different auth and access pattern. 

In summary to support deploying to cloudrun on anthos we would need to implement a new backend for the GCP provider that supported Kubernetes. Since there are already Terraform providers which allow for managing Kubernetes clusters that can be used in coordination with the GCP provider to fully manage a GKE cluster. I would look into the Helm and Kubernetes Terraform providers and if you have a Cloud Run use case that isn't supported by them we can figure out which provider will need a feature added.
